src/parser.rs

```diff
@@ -1,726 +1,451 @@
-            't' => Ok(('\t', 2, true)),
-            'u' => {
-                if chars.next() != Some('{') {
-                    return Err(ParserError::InvalidEscape(
-                        span_str.chars().skip(1).take(2).collect(),
-                    ));
-                }
-                let mut nibbles = chars.clone().map(|ch| ch.to_digit(16));
-                let mut num_nibbles = 0;
-                let mut value: u32 = 0;
-                while let Some(Some(nibble)) = nibbles.next() {
-                    num_nibbles += 1;
-                    value <<= 4;
-                    value |= nibble;
-                    if value > 0x10FFFF {
-                        return Err(ParserError::InvalidEscape(
-                            span_str.chars().skip(1).take(2 + num_nibbles + 1).collect(),
-                        ));
+            match self.advance()? {
+                Token::Comma => {
+                    if self.next_is(end_token) {
+                        self.advance()?;
+                        break;
                     }
                 }
-                if chars.nth(num_nibbles) != Some('}') {
-                    return Err(ParserError::InvalidEscape(
-                        span_str.chars().skip(1).take(2 + num_nibbles + 1).collect(),
-                    ));
-                }
-                match value.try_into() {
-                    Ok(ch) => Ok((ch, 3 + num_nibbles + 1, true)),
-                    Err(_) => Err(ParserError::InvalidEscape(
-                        span_str.chars().skip(1).take(2 + num_nibbles + 1).collect(),
-                    )),
+                _ => {
+                    self.expect_token(end_token)?;
+                    break;
                 }
             }
-            other => Err(ParserError::InvalidEscape(other.to_string())),
         }
+        Ok(nodes)
     }
 
-    fn expect_name<T: Borrow<str>>(
-        &mut self,
-        names: impl IntoIterator<Item = T>,
-    ) -> Result<&str, ParserError> {
-        let name = match self.parse_name() {
-            Ok(name) => name,
-            Err(ParserError::UnexpectedToken { got: None, .. }) => "",
-            Err(other) => return Err(other),
-        };
-        let names = names.into_iter().collect::<Vec<_>>();
-        if names.iter().any(|n| n.borrow() == name) {
-            Ok(name)
-        } else {
-            Err(ParserError::unexpected_name(
-                names.iter().map(|n| n.borrow().to_string()),
-                name,
-            ))
-        }
+    fn leaf_node(&self, ty: NodeType) -> Node {
+        Node::new(ty, self.span(), [])
     }
+}
 
-    fn parse_name(&mut self) -> Result<&str, ParserError> {
-        let span = self.expect(Token::Name)?;
-        let name = self.tokens.get_span(span);
-        Ok(name.strip_prefix('%').unwrap_or(name))
+pub(crate) fn flattenable(kind: WasmTypeKind) -> bool {
+    use WasmTypeKind::*;
+    !matches!(kind, Variant | Enum | Option | Result)
```

Comment:
With some/none/ok/err being keywords, is the plan still to make variant/enum flattenable? Or further, to make option-in-result or result-in-option flattenable? If so it'd be fine to do a TODO comment here for now.

Reply:
Updated to allow inner variant/enum. I haven't thought through mixed option/result flattening yet so added a note as well.